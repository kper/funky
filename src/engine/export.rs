use wasm_parser::core::{ExternalKindType, ExportEntry};

#[derive(Debug, Clone, PartialEq)]
pub struct ExportInstance {
    pub name: String,
    pub value: ExternalKindType,
}

impl From<&ExportEntry> for ExportInstance {
    fn from(state: &ExportEntry) -> ExportInstance {
        ExportInstance {
            name: state.name.clone(),
            value: state.kind,
        }
    }
}
