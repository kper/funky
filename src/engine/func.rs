use wasm_parser::core::{FunctionBody, FunctionSignature};

#[derive(Debug, Clone)]
pub struct FuncInstance {
    //FIXME Add HostFunc
    pub ty: FunctionSignature,
    pub code: FunctionBody,
}
