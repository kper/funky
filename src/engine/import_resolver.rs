use crate::engine::{GlobalInstance, TableInstance};
use anyhow::{anyhow, Result};
use std::collections::HashMap;
//use wasm_parser::core::ImportEntry;

pub type Imports = Vec<Import>;

type Module = String;
type Name = String;

//TODO add more types
#[derive(Debug)]
pub enum Import {
    Global(Module, Name, GlobalInstance),
    Table(Module, Name, TableInstance),
}

/// Private lookup table for
/// modules and names
#[derive(Debug, Default)]
struct LookupTable {
    modules: HashMap<(Module, Name), Import>,
}

impl LookupTable {
    pub fn lookup(&self, module: impl Into<String>, name: impl Into<String>) -> Option<&Import> {
        if let Some(m) = self.modules.get(&(module.into(), name.into())) {
            return Some(m);
        }

        None
    }
}

/// The `ImportResolver` matches the `wasm_parser::core::ImportDesc`
/// with the actual given imports.
#[derive(Debug)]
pub(crate) struct ImportResolver {
    imports: LookupTable,
}

impl ImportResolver {
    pub fn new() -> Self {
        Self {
            imports: LookupTable::default(),
        }
    }

    /// Get the imported global by module and name
    pub fn resolve_global(&self, module: &String, name: &String) -> Result<GlobalInstance> {
        debug!("resolve global {} {}", module, name);

        if let Some(Import::Global(_, _, instance)) = self.imports.lookup(module, name) {
            return Ok(instance.clone());
        }

        Err(anyhow!("Cannot find global for {} {}", module, name))
    }

    /// Get the imported table by module and name
    pub fn resolve_table(&self, module: &String, name: &String) -> Result<TableInstance> {
        debug!("resolve table {} {}", module, name);

        if let Some(Import::Table(_, _, instance)) = self.imports.lookup(module, name) {
            return Ok(instance.clone());
        }

        Err(anyhow!("Cannot find table for {} {}", module, name))
    }

    pub fn inject_global(
        &mut self,
        module: Module,
        name: Name,
        instance: &GlobalInstance,
    ) -> Result<()> {
        self.imports.modules.insert(
            (module.clone(), name.clone()),
            Import::Global(module, name, instance.clone()),
        );

        Ok(())
    }

    pub fn inject_table(
        &mut self,
        module: Module,
        name: Name,
        instance: &TableInstance,
    ) -> Result<()> {
        self.imports.modules.insert(
            (module.clone(), name.clone()),
            Import::Table(module, name, instance.clone()),
        );

        Ok(())
    }
}
