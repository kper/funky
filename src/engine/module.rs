use crate::engine::export::ExportInstance;
use anyhow::{anyhow, Result};
use wasm_parser::core::*;
use wasm_parser::Module;

#[derive(Debug, Clone)]
pub struct ModuleInstance {
    pub start: u32,
    pub code: Vec<FunctionBody>,
    pub fn_types: Vec<FunctionSignature>,
    pub funcaddrs: Vec<FuncAddr>,
    pub tableaddrs: Vec<TableIdx>,
    pub memaddrs: Vec<MemoryIdx>,
    pub globaladdrs: Vec<GlobalAddr>,
    pub exports: Vec<ExportInstance>,
}

impl ModuleInstance {
    pub fn new(m: &Module) -> Self {
        let mut mi = ModuleInstance {
            start: 0,
            code: Vec::new(),
            fn_types: Vec::new(),
            funcaddrs: Vec::new(),
            tableaddrs: Vec::new(),
            memaddrs: Vec::new(),
            globaladdrs: Vec::new(),
            exports: Vec::new(),
        };
        for section in m.sections.iter() {
            match section {
                Section::Code(CodeSection { entries: x }) => {
                    mi.code = x.clone();
                }
                Section::Type(TypeSection { entries: x }) => {
                    mi.fn_types = x.clone();
                }
                _ => {}
            }
        }

        mi
    }

    /// Adding a new function type.
    /// We need this function to test blocks, with multiple
    /// return values.
    pub(crate) fn add_func_type(&mut self, r: Vec<ValueType>) -> usize {
        let instance = FunctionSignature {
            param_types: vec![],
            return_types: r,
        };

        self.fn_types.push(instance);

        self.fn_types.len() - 1
    }

    /// Looking up the function's address in the store by given function's module address.
    pub fn lookup_function_addr(&self, function_module_addr: FuncIdx) -> Result<FuncAddr> {
        debug!(
            "Looking up store's func addr for idx {}",
            function_module_addr
        );

        Ok(self
            .funcaddrs
            .get(function_module_addr as usize)
            .ok_or_else(|| anyhow!("Cannot lookup funcaddr for id {:?}", function_module_addr))?
            .clone())
    }
}
