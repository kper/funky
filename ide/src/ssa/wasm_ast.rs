#![allow(dead_code)]
use anyhow::{bail, Result};
/// This module is responsible to parse
/// the webassembly AST to an IR
use funky::engine::func::FuncInstance;
use funky::engine::Engine;
use log::debug;
use std::fmt::Write;
use wasm_parser::core::Instruction::*;
use wasm_parser::core::*;
use crate::symbol_table::SymbolTable;

use std::collections::HashMap;

#[derive(Debug)]
pub struct IR<'a> {
    //functions: Vec<Function>,
    buffer: String,
    //counter: Counter,
    symbol_table: SymbolTable,
    block_counter: Counter,
    function_counter: Counter,
    functions: Vec<Function>,
    engine: &'a Engine,
}

#[derive(Debug)]
struct Function {
    name: String,
    locals: HashMap<usize, usize>,  //key to register
    globals: HashMap<usize, usize>, //key to register
}

#[derive(Debug)]
struct Block {
    name: usize,
    is_loop: bool,
}

#[derive(Debug)]
struct JumpList {
    blocks: Vec<Block>,
}

#[derive(Debug, Default)]
struct Counter {
    counter: usize,
}

impl Counter {
    pub fn peek(&self) -> usize {
        self.counter
    }

    pub fn get(&mut self) -> usize {
        let counter = self.counter.clone();
        self.counter += 1;
        counter
    }

    pub fn peek_next(&self) -> usize {
        self.counter + 1
    }
}


impl<'a> IR<'a> {
    pub fn new(engine: &'a Engine) -> Self {
        Self {
            engine,
            buffer: String::new(),
            //counter: Counter::default(),
            symbol_table: SymbolTable::default(),
            block_counter: Counter::default(),
            function_counter: Counter::default(),
            functions: Vec::new(),
        }
    }

    pub fn buffer(&self) -> String {
        self.buffer.clone()
    }

    pub fn visit(&mut self) -> Result<()> {
        for function in self.engine.store.funcs.iter() {
            self.visit_function(function)?;
        }

        Ok(())
    }

    fn visit_function(&mut self, inst: &FuncInstance) -> Result<()> {
        let name = format!("{}", self.function_counter.get());
        writeln!(self.buffer, "define {} {{", name).unwrap();

        let mut function = Function {
            name,
            locals: HashMap::new(),
            globals: HashMap::new(),
        };

        for (i, _) in inst.ty.param_types.iter().enumerate() {
            function.locals.insert(i, self.symbol_table.new_var()?);
        }

        self.functions.push(function);
        let func_index = self.functions.len() - 1;

        debug!("code {:#?}", inst.code);
        self.visit_body(&inst.code, func_index, inst.ty.return_types.len())?;

        writeln!(self.buffer, "}};").unwrap();

        Ok(())
    }

    fn visit_body(&mut self, body: &FunctionBody, function_index: usize, return_count: usize) -> Result<()> {
        let code = &body.code;

        let name = self.block_counter.get();
        let then_name = self.block_counter.get();

        let block = Block {
            name: name.clone(),
            is_loop: false,
        };

        let mut blocks = vec![block];

        writeln!(self.buffer, "BLOCK {}", name).unwrap();

        self.visit_instruction_wrapper(code, function_index, &mut blocks, return_count)?;

        writeln!(self.buffer, "BLOCK {}", then_name).unwrap();

        Ok(())
    }

    /// If the execution exits a block (with no jump),
    /// then kill all variables which are not returned.
    fn exit_block(&mut self, block_ty: &BlockType, old_state: usize) -> Result<()> {
        let arity = self.engine.get_return_count_block(block_ty)?;

        for reg in self.symbol_table.vars.iter_mut().rev().skip(arity as usize) {
            if reg.val() <= old_state {
                break;
            }

            if !reg.is_killed {
                reg.is_killed = true;
                writeln!(self.buffer, "KILL %{}", reg.val()).unwrap();
            }
        }

        Ok(())
    }

    fn visit_instruction_wrapper(
        &mut self,
        code: &[InstructionWrapper],
        function_index: usize,
        blocks: &mut Vec<Block>,
        return_arity: usize,
    ) -> Result<()> {
        debug!("Visiting instruction wrapper");

        //let mut str_block = String::new();
        //writeln!(str_block, "BLOCK {}", self.block_counter.get()).unwrap();

        let blocks_len = blocks.len();

        for instr in code.iter() {
            debug!("Instruction {}", instr.get_instruction());

            match instr.get_instruction() {
                OP_BLOCK(ty, code) => {
                    let name = self.block_counter.get();
                    let then_name = self.block_counter.get();

                    let block = Block {
                        name: name.clone(),
                        is_loop: false,
                    };

                    let _tblock = Block {
                        name: then_name.clone(),
                        is_loop: false,
                    };

                    blocks.push(block);

                    writeln!(self.buffer, "BLOCK {}", name.clone()).unwrap();

                    // If the block exits, then kill all variable to `current_reg`
                    let current_reg = self.symbol_table.peek()?;

                    let arity = self.engine.get_return_count_block(ty)?;
                    self.visit_instruction_wrapper(
                        code.get_instructions(),
                        function_index,
                        blocks,
                        arity as usize
                    )?;

                    self.exit_block(ty, current_reg)?;

                    blocks.pop();

                    writeln!(
                        self.buffer,
                        "GOTO {} // BLOCK ended for {}",
                        then_name,
                        name.clone()
                    )
                    .unwrap();
                    writeln!(
                        self.buffer,
                        "BLOCK {} // THEN block for {}",
                        then_name, name
                    )
                    .unwrap();
                }
                /*
                OP_LOOP(ty, code) => {
                    let name = self.block_counter.get();
                    let then_name = self.block_counter.get();

                    let block = Block {
                        name: name.clone(),
                        is_loop: true,
                    };

                    let _tblock = Block {
                        name: then_name.clone(),
                        is_loop: false,
                    };

                    blocks.push(block);

                    writeln!(self.buffer, "BLOCK {} // LOOP", name.clone()).unwrap();

                    let arity = self.engine.get_return_count_block(ty)?;
                    self.visit_instruction_wrapper(
                        code.get_instructions(),
                        function_index,
                        blocks,
                        arity as usize,
                    )?;

                    blocks.pop();

                    writeln!(
                        self.buffer,
                        "GOTO {} // BLOCK ended for {}",
                        then_name,
                        name.clone()
                    )
                    .unwrap();
                    writeln!(
                        self.buffer,
                        "BLOCK {} // THEN block for {}",
                        then_name, name
                    )
                    .unwrap();
                }
                OP_IF(ty, code) => {
                    let name = self.block_counter.get();
                    let then_name = self.block_counter.get();

                    let block = Block {
                        name: name.clone(),
                        is_loop: false,
                    };

                    let _tblock = Block {
                        name: then_name.clone(),
                        is_loop: false,
                    };

                    blocks.push(block);

                    writeln!(
                        self.buffer,
                        "IF %{} THEN GOTO {} ELSE GOTO {}",
                        self.symbol_table.peek()?,
                        name.clone(),
                        then_name.clone()
                    )
                    .unwrap();
                    writeln!(self.buffer, "BLOCK {} ", name.clone()).unwrap();

                    let arity = self.engine.get_return_count_block(ty)?;
                    self.visit_instruction_wrapper(
                        code.get_instructions(),
                        function_index,
                        blocks,
                        arity as usize,
                    )?;

                    blocks.pop();

                    writeln!(
                        self.buffer,
                        "GOTO {} // BLOCK ended for {}",
                        then_name,
                        name.clone()
                    )
                    .unwrap();
                    writeln!(
                        self.buffer,
                        "BLOCK {} // THEN block for {}",
                        then_name, name
                    )
                    .unwrap();
                }
                OP_IF_AND_ELSE(ty, code1, code2) => {
                    let name = self.block_counter.get();
                    let then_name = self.block_counter.get();
                    let done_name = self.block_counter.get();

                    let block = Block {
                        name: name.clone(),
                        is_loop: false,
                    };

                    let tblock = Block {
                        name: then_name.clone(),
                        is_loop: false,
                    };

                    let _done_block = Block {
                        name: done_name.clone(),
                        is_loop: false,
                    };

                    blocks.push(block);

                    writeln!(
                        self.buffer,
                        "IF %{} THEN GOTO {} ELSE GOTO {}",
                        self.symbol_table.peek()?,
                        name.clone(),
                        then_name.clone()
                    )
                    .unwrap();
                    writeln!(self.buffer, "BLOCK {} ", name.clone()).unwrap();

                    let arity = self.engine.get_return_count_block(ty)?;
                    self.visit_instruction_wrapper(
                        code1.get_instructions(),
                        function_index,
                        blocks,
                        arity as usize,
                    )?;

                    writeln!(
                        self.buffer,
                        "GOTO {} // BLOCK ended for {}",
                        done_name,
                        name.clone()
                    )
                    .unwrap();
                    writeln!(
                        self.buffer,
                        "BLOCK {} // THEN block for {}",
                        then_name, name
                    )
                    .unwrap();

                    blocks.pop();

                    blocks.push(tblock);

                    self.visit_instruction_wrapper(
                        code2.get_instructions(),
                        function_index,
                        blocks,
                        arity as usize,
                    )?;

                    blocks.pop();

                    writeln!(
                        self.buffer,
                        "GOTO {} // BLOCK ended for {}",
                        done_name,
                        name.clone()
                    )
                    .unwrap();

                    writeln!(
                        self.buffer,
                        "BLOCK {} // THEN block for {}",
                        done_name, then_name
                    )
                    .unwrap();
                }*/
                OP_BR(label) => {
                    let jmp_index = blocks_len - 1 - *label as usize;

                    let block = blocks.get(jmp_index).unwrap();

                    if block.is_loop {
                        writeln!(self.buffer, "GOTO {}", block.name).unwrap();
                    } else {
                        writeln!(self.buffer, "GOTO {}", block.name + 1).unwrap();
                    }
                }
                OP_BR_IF(label) => {
                    let jmp_index = blocks_len - 1 - *label as usize;

                    let block = blocks.get(jmp_index).unwrap();

                    if block.is_loop {
                        writeln!(
                            self.buffer,
                            "IF %{} THEN GOTO {}",
                            self.symbol_table.peek()?,
                            block.name
                        )
                        .unwrap();
                    } else {
                        writeln!(
                            self.buffer,
                            "IF %{} THEN GOTO {}",
                            self.symbol_table.peek()?,
                            block.name + 1
                        )
                        .unwrap();
                    }
                }
                OP_BR_TABLE(labels, else_lb) => {
                    let indices: Vec<_> = labels
                        .iter()
                        .map(|x| {
                            let i = blocks_len - 1 - *x as usize;

                            let block = blocks.get(i).unwrap();
                            if block.is_loop {
                                block.name
                            } else {
                                block.name + 1
                            }
                        })
                        .map(|x| format!("{}", x))
                        .collect();

                    let jmp_index = blocks_len - 1 - *else_lb as usize;
                    let block = blocks.get(jmp_index).unwrap();

                    let jmp_index = match block.is_loop {
                        true => block.name,
                        false => block.name + 1,
                    };

                    writeln!(
                        self.buffer,
                        "BR TABLE GOTO {} ELSE GOTO {}",
                        indices.join(" "),
                        jmp_index
                    )
                    .unwrap();
                }
                OP_LOCAL_GET(index) => {
                    writeln!(
                        self.buffer,
                        "%{} = %{}",
                        self.symbol_table.new_var()?,
                        self.functions
                            .get(function_index)
                            .unwrap()
                            .locals
                            .get(&(*index as usize))
                            .unwrap()
                    )
                    .unwrap();
                }
                OP_LOCAL_SET(index) => {
                    writeln!(
                        self.buffer,
                        "%{} = %{}",
                        self.functions
                            .get(function_index)
                            .unwrap()
                            .locals
                            .get(&(*index as usize))
                            .unwrap(),
                        self.symbol_table.peek()?
                    )
                    .unwrap();
                }
                OP_LOCAL_TEE(index) => {
                    println!("{:?}", self.symbol_table);
                    let peek = self.symbol_table.peek()?;
                    // Push only once because the old still lives
                    writeln!(self.buffer, "%{} = %{}", self.symbol_table.new_var()?, peek).unwrap();
                    writeln!(
                        self.buffer,
                        "%{} = %{}",
                        self.functions
                            .get(function_index)
                            .unwrap()
                            .locals
                            .get(&(*index as usize))
                            .unwrap(),
                        peek
                    )
                    .unwrap();
                }
                OP_CALL(func) => {
                    let addr = self.engine.module.lookup_function_addr(*func)?;
                    let instance = self.engine.store.get_func_instance(&addr)?;

                    let num_params = instance.ty.param_types.len();

                    let mut param_regs = Vec::new();

                    for i in 0..num_params {
                        param_regs.push(format!("{}", self.symbol_table.peek_offset(i)?));
                    }

                    writeln!(self.buffer, "CALL {}({})", func, param_regs.join(",")).unwrap();
                }
                OP_I32_CONST(a) => {
                    writeln!(self.buffer, "%{} = {}", self.symbol_table.new_var()?, a).unwrap();
                }
                OP_I64_CONST(a) => {
                    writeln!(self.buffer, "%{} = {}", self.symbol_table.new_var()?, a).unwrap();
                }
                OP_F32_CONST(a) => {
                    writeln!(self.buffer, "%{} = {}", self.symbol_table.new_var()?, a).unwrap();
                }
                OP_F64_CONST(a) => {
                    writeln!(self.buffer, "%{} = {}", self.symbol_table.new_var()?, a).unwrap();
                }
                OP_RETURN => {
                    let mut regs = Vec::new(); 

                    for i in 0..return_arity {
                        regs.push(format!("%{}", self.symbol_table.peek_offset(i)?));
                    }

                    writeln!(self.buffer, "return {}", regs.join(" ")).unwrap();
                }
                OP_I32_CLZ
                | OP_I32_CTZ
                | OP_I32_POPCNT
                | OP_I64_CLZ
                | OP_I64_CTZ
                | OP_I64_POPCNT
                | OP_F32_ABS
                | OP_F32_NEG
                | OP_F32_CEIL
                | OP_F32_FLOOR
                | OP_F32_TRUNC
                | OP_F32_NEAREST
                | OP_F32_SQRT
                | OP_F64_ABS
                | OP_F64_NEG
                | OP_F64_CEIL
                | OP_F64_FLOOR
                | OP_F64_TRUNC
                | OP_F64_NEAREST
                | OP_F64_SQRT
                | OP_I32_WRAP_I64
                | OP_I32_TRUNC_F32_S
                | OP_I32_TRUNC_F32_U
                | OP_I32_TRUNC_F64_S
                | OP_I32_TRUNC_F64_U
                | OP_I64_EXTEND_I32_U
                | OP_I64_EXTEND_I32_S
                | OP_I64_TRUNC_F32_S
                | OP_I64_TRUNC_F32_U
                | OP_I64_TRUNC_F64_S
                | OP_I64_TRUNC_F64_U
                | OP_F32_CONVERT_I32_S
                | OP_F32_CONVERT_I32_U
                | OP_F32_CONVERT_I64_S
                | OP_F32_CONVERT_I64_U
                | OP_F32_DEMOTE_F64
                | OP_F64_CONVERT_I32_S
                | OP_F64_CONVERT_I32_U
                | OP_F64_CONVERT_I64_S
                | OP_F64_CONVERT_I64_U
                | OP_F64_PROMOTE_F32
                | OP_I32_REINTERPRET_F32
                | OP_I64_REINTERPRET_F64
                | OP_F32_REINTERPRET_I32
                | OP_F64_REINTERPRET_I64
                | OP_I32_EXTEND8_S
                | OP_I32_EXTEND16_S
                | OP_I64_EXTEND8_S
                | OP_I64_EXTEND16_S
                | OP_I64_EXTEND32_S
                | OP_I32_TRUNC_SAT_F32_S
                | OP_I32_TRUNC_SAT_F32_U
                | OP_I32_TRUNC_SAT_F64_S
                | OP_I32_TRUNC_SAT_F64_U
                | OP_I64_TRUNC_SAT_F32_S
                | OP_I64_TRUNC_SAT_F32_U
                | OP_I64_TRUNC_SAT_F64_S
                | OP_I64_TRUNC_SAT_F64_U => {
                    writeln!(
                        self.buffer,
                        "%{} = {} %{}",
                        self.symbol_table.new_var()?,
                        "op",
                        self.symbol_table.peek_offset(2)?, // TODO Check why `2`
                    )
                    .unwrap();
                }
                OP_I32_ADD | OP_I32_SUB | OP_I32_MUL | OP_I32_DIV_S | OP_I32_DIV_U
                | OP_I32_REM_S | OP_I32_REM_U | OP_I32_AND | OP_I32_OR | OP_I32_XOR
                | OP_I32_SHL | OP_I32_SHR_S | OP_I32_SHR_U | OP_I32_ROTL | OP_I32_ROTR
                | OP_I64_ADD | OP_I64_SUB | OP_I64_MUL | OP_I64_DIV_S | OP_I64_DIV_U
                | OP_I64_REM_S | OP_I64_REM_U | OP_I64_AND | OP_I64_OR | OP_I64_XOR
                | OP_I64_SHL | OP_I64_SHR_S | OP_I64_SHR_U | OP_I64_ROTL | OP_I64_ROTR
                | OP_I32_EQZ | OP_I32_EQ | OP_I32_NE | OP_I32_LT_S | OP_I32_LT_U | OP_I32_GT_S
                | OP_I32_GT_U | OP_I32_LE_S | OP_I32_LE_U | OP_I32_GE_S | OP_I32_GE_U
                | OP_I64_EQZ | OP_I64_EQ | OP_I64_NE | OP_I64_LT_S | OP_I64_LT_U | OP_I64_GT_S
                | OP_I64_GT_U | OP_I64_LE_S | OP_I64_LE_U | OP_I64_GE_S | OP_I64_GE_U
                | OP_F32_EQ | OP_F32_NE | OP_F32_LT | OP_F32_GT | OP_F32_LE | OP_F32_GE
                | OP_F64_EQ | OP_F64_NE | OP_F64_LT | OP_F64_GT | OP_F64_LE | OP_F64_GE
                | OP_F32_ADD | OP_F32_SUB | OP_F32_MUL | OP_F32_DIV | OP_F64_ADD | OP_F64_SUB
                | OP_F64_MUL | OP_F64_DIV | OP_F32_MIN | OP_F32_MAX | OP_F32_COPYSIGN
                | OP_F64_MIN | OP_F64_MAX | OP_F64_COPYSIGN => {
                    writeln!(
                        self.buffer,
                        "%{} = %{} {} %{}",
                        self.symbol_table.new_var()?,
                        self.symbol_table.peek_offset(1)?,
                        "op",
                        self.symbol_table.peek_offset(2)?
                    )
                    .unwrap();
                }
                _ => {
                    writeln!(self.buffer, "{}", instr.get_instruction()).unwrap();
                }
            }
        }

        Ok(())

        /*
        if !jumped {
            writeln!(str_block, "GOTO {}", self.block_counter.peek()).unwrap();
        }*/

        //writeln!(str_block, "BLOCK {}", self.block_counter.get()).unwrap();

        //writeln!(self.buffer, "{}", str_block).unwrap();
    }
}
