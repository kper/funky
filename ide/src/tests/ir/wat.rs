use crate::ir::wasm_ast::IR;
use funky::engine::module::ModuleInstance;
use funky::engine::*;
use insta::assert_snapshot;
use std::fs::File;
use std::io::{self, Write};
use validation::validate;
use wasm_parser::{parse, read_wasm};

macro_rules! wat {
    ($name:expr, $input:expr) => {{
        use std::process::Command;
        use tempfile::tempdir;

        let tmp_dir = tempdir().unwrap();
        let file_path = tmp_dir.path().join("file.wat");

        {
            let mut tmp_file = File::create(file_path.clone()).unwrap();
            writeln!(tmp_file, $input).unwrap();
            tmp_file.flush().unwrap();

            println!("tmp_file {}", $input);
        }

        let output = Command::new("wat2wasm")
            .arg(file_path)
            .arg("-o")
            .arg(tmp_dir.path().join("file.wasm").to_str().unwrap())
            .output()
            .expect("Command failed");

        io::stdout().write_all(&output.stdout).unwrap();
        io::stderr().write_all(&output.stderr).unwrap();

        assert!(output.status.success());

        let file_path = tmp_dir.path().join("file.wasm");

        // Read it
        let file = read_wasm!(file_path);
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let imports = Vec::new();

        let instance = ModuleInstance::new(&module);
        let engine = Engine::new(
            instance,
            &module,
            Box::new(funky::debugger::RelativeProgramCounter::default()),
            &imports,
        )
        .unwrap();

        let mut ir = IR::new();

        ir.visit(&engine).unwrap();

        assert_snapshot!($name, format!("{}", ir.buffer()));

        ir
    }};
}

#[test]
fn test_empty_function() {
    wat!("test_empty_function", "(module (func))");
}

#[test]
fn test_simple_block() {
    wat!(
        "test_simple_block",
        "(module (func (result i32) (block (result i32) i32.const 1)))"
    );
}

#[test]
fn test_nested_block() {
    wat!(
        "test_nested_block",
        "(module (func (result i32) (block (result i32) (block (result i32) i32.const 1))))"
    );
}

#[test]
fn test_simple_loop() {
    wat!("test_simple_loop", "(module (func (result i32) (block (result i32) (loop (result i32) (i32.const 1) (br 1)))))");
}

#[test]
fn test_simple_if() {
    wat!("test_if", "(module (func (if (i32.const 1) (then nop))))");
}

#[test]
fn test_simple_if_and_else() {
    wat!(
        "test_if_else",
        "(module (func (if (i32.const 1) (then nop) (else nop))))"
    );
}

#[test]
fn test_simple_br_if() {
    wat!("test_br_if", "(module (func (i32.const 1) (br_if 0)))");
}

#[test]
fn test_simple_br_table() {
    wat!(
        "test_br_table",
        "(module (func (param i32) (result i32)
    (block
      (block
        (br_table 1 0 (local.get 0))
        (return (i32.const 21))
      )
      (return (i32.const 20))
    )
    (i32.const 22)
  ))"
    );
}

#[test]
fn test_nested_block_value() {
    wat!(
        "test_nested_block_value",
        "(module
     (func (export \"nested-block-value\") (param i32) (result i32)
    (block (result i32)
      (drop (i32.const -1))
      (i32.add
        (i32.const 1)
        (block (result i32)
          (i32.add
            (i32.const 2)
            (block (result i32)
              (drop (i32.const 4))
              (i32.add
                (i32.const 8)
                (br_table 0 1 2 (i32.const 16) (local.get 0))
              )
            )
          )
        )))))"
    );
}

#[test]
fn test_local_tee() {
    wat!(
        "local_tee",
        "(module
    (func (param i32) (result i32) (i32.const 1) (local.tee 0)) 
  )"
    );
}

#[test]
fn test_call() {
    wat!(
        "test_call",
        "(module
    (func $fib (export \"fib\") (param i64) (result i64)
    (if (result i64) (i64.le_u (local.get 0) (i64.const 1))
      (then (i64.const 1))
      (else
        (i64.add
          (call $fib (i64.sub (local.get 0) (i64.const 2)))
          (call $fib (i64.sub (local.get 0) (i64.const 1)))
        )
      )
    )
  )

  (func $even (export \"even\") (param i64) (result i32)
    (if (result i32) (i64.eqz (local.get 0))
      (then (i32.const 44))
      (else (call $odd (i64.sub (local.get 0) (i64.const 1))))
    )
  )
  (func $odd (export \"odd\") (param i64) (result i32)
    (if (result i32) (i64.eqz (local.get 0))
      (then (i32.const 99))
      (else (call $even (i64.sub (local.get 0) (i64.const 1))))
    )
  )
    )"
    );
}

#[test]
fn test_phi() {
    wat!(
        "test_phi",
        "(module
	        (func (export \"singular\") (param i32) (result i32)
	          (i32.const 3)	
            (if (result i32) (local.get 0) (then (i32.const 1)) (else (i32.const 2)))
	          (i32.add)
          ))"
    );
}
#[test]
fn test_phi2() {
    wat!(
        "test_phi2",
        "(module
	        (func (export \"singular\") (param i32) (result i32)
	          (i32.const 4)	
            (if (result i32 i32) (local.get 0) (then (i32.const 1) (i32.const 1)) (else (i32.const 2) (i32.const 2)))
	          (i32.add)
	          (i32.add)
          ))"
    );
}

#[test]
fn test_fac() {
    wat!(
        "test_fac",
        "(module
          (func (export \"fac-rec\") (param i64) (result i64)
            (if (result i64) (i64.eq (local.get 0) (i64.const 0))
              (then (i64.const 1))
              (else
                (i64.mul (local.get 0) (call 0 (i64.sub (local.get 0) (i64.const 1))))
              )
            )
          ))"
    );
}

#[test]
fn test_fac_reversed() {
    wat!(
        "test_fac_reversed",
        "(module
          (func (export \"fac-rec\") (param i64) (result i64)
            (if (result i64) (i64.eq (local.get 0) (i64.const 0))
              (then (i64.mul (local.get 0) (call 0 (i64.sub (local.get 0) (i64.const 1)))))
              (else
                (i64.const 1)
              )
            )
          ))"
    );
}

#[test]
fn test_call_indirect() {
    wat!(
        "test_call_indirect",
        "(module
          (type $out-i32 (func (result i32)))
          (type $out-i64 (func (result i64)))
          (type $out-f32 (func (result f32)))
          (type $out-f64 (func (result f64)))
          (func $const-i32 (type $out-i32) (i32.const 0x132))
          (func $const-i64 (type $out-i64) (i64.const 0x164))
          (func $const-f32 (type $out-f32) (f32.const 0xf32))
          (func $const-f64 (type $out-f64) (f64.const 0xf64))
          (table funcref
          (elem
            $const-i32 $const-i64 $const-f32 $const-f64  ;; 0..3
          ))
          (func (export \"type-i32\") (result i32)
            (call_indirect (type $out-i32) (i32.const 0))
          )
          (func (export \"type-i64\") (result i64)
            (call_indirect (type $out-i64) (i32.const 1))
          )
          (func (export \"type-f32\") (result f32)
            (call_indirect (type $out-f32) (i32.const 2))
          )
          (func (export \"type-f64\") (result f64)
            (call_indirect (type $out-f64) (i32.const 3))
          ))"
    );
}

#[test]
fn test_call_indirect_with_param() {
    wat!(
        "test_call_indirect_with_param",
        "(module
          
        (type $over-i32 (func (param i32) (result i32)))
        (type $over-i64 (func (param i64) (result i64)))
        (type $over-f32 (func (param f32) (result f32)))
        (type $over-f64 (func (param f64) (result f64)))

        (func $id-i32 (type $over-i32) (local.get 0))
        (func $id-i64 (type $over-i64) (local.get 0))
        (func $id-f32 (type $over-f32) (local.get 0))
        (func $id-f64 (type $over-f64) (local.get 0))

        (table funcref
          (elem
            $id-i32 $id-i64 $id-f32 $id-f64
          )
        )

        (func (export \"type-first-i32\") (result i32)
          (call_indirect (type $over-i32) (i32.const 32) (i32.const 0))
        )
        (func (export \"type-first-i64\") (result i64)
          (call_indirect (type $over-i64) (i64.const 64) (i32.const 1))
        )
        (func (export \"type-first-f32\") (result f32)
          (call_indirect (type $over-f32) (f32.const 1.32) (i32.const 2))
        )
        (func (export \"type-first-f64\") (result f64)
          (call_indirect (type $over-f64) (f64.const 1.64) (i32.const 3))
        ))"
    );
}


#[test]
fn test_global_get() {
    wat!(
        "test_global_get",
        "(module

            (global $a i32 (i32.const -2))
            (global (;1;) f32 (f32.const -3))
            (global (;2;) f64 (f64.const -4))
            (global $b i64 (i64.const -5))

            (global $x (mut i32) (i32.const -12))
            (global (;5;) (mut f32) (f32.const -13))
            (global (;6;) (mut f64) (f64.const -14))
            (global $y (mut i64) (i64.const -15))

            (func (export \"get-a\") (result i32) (global.get $a))
            (func (export \"get-b\") (result i64) (global.get $b))
            (func (export \"get-x\") (result i32) (global.get $x))
            (func (export \"get-y\") (result i64) (global.get $y))
        )"
    );
}

#[test]
fn test_global_set() {
    wat!(
        "test_global_set",
        "(module

            (global $a i32 (i32.const -2))
            (global (;1;) f32 (f32.const -3))
            (global (;2;) f64 (f64.const -4))
            (global $b i64 (i64.const -5))

            (global $x (mut i32) (i32.const -12))
            (global (;5;) (mut f32) (f32.const -13))
            (global (;6;) (mut f64) (f64.const -14))
            (global $y (mut i64) (i64.const -15))

            (func (export \"set-x\") (param i32) (global.set $x (local.get 0)))
            (func (export \"set-y\") (param i64) (global.set $y (local.get 0)))
        )"
    );
}