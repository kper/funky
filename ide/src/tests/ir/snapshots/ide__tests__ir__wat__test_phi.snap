---
source: src/tests/ir/wat.rs
expression: "format!(\"{}\", ir . buffer())"

---
define 0 (param %0) (result 1) (define %0 %1 %2 %3 %4 %5 %6) {
BLOCK 0
%1 = 3
%2 = %0
IF %2 THEN GOTO 1 ELSE GOTO 2
BLOCK 1 
%3 = 1
GOTO 3
BLOCK 2
%4 = 2
GOTO 3
BLOCK 3
%5 = phi %3 %4
%6 = %5 op %2
RETURN %6;
};

